dimod/include/dimod/iterators.h

```diff
@@ -0,0 +1,124 @@
+// Copyright 2022 D-Wave Systems Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#pragma once
+
+namespace dimod {
+namespace iterators {
+
+template <class Iter, class T = typename std::iterator_traits<Iter>::value_type::element_type>
```

Comment:
Will it work correctly if T is not the default one you are setting. Are there corner cases for that ?